tests/test_waiting.py

```diff
@@ -104,47 +104,58 @@ def test_wait_epoll_bad(pgconn):
     assert res.status == ExecStatus.TUPLES_OK
 
 
+@pytest.fixture
+def wait_async(anyio_backend_name):
+    return {
+        "asyncio": waiting.wait_asyncio,
+        "trio": waiting.wait_anyio,
+    }[anyio_backend_name]
```

Comment:
This way the test depends on trio/anyio. It means that we are not able to verify anymore that psycopg works if trio is not installed.

I think that we should run the trio tests in a separate environment that would install `psycopg[trio]` and would test the anyio objects, whereas in the base tests we would test taht AnyIOConnection etc. "not work as expected".

Reply:
The test suite needs `anyio`'s pytest plugin, per first commit (and discussions above) where it replaces `pytest-asyncio`. But since the plugin comes along with the library, the test suite actually depends on the whole `anyio` package.

Maybe we could make the `trio` dependency optional in the test suite (to be caught in `anyio_backend` fixture, I'd suppose). But I'm not sure I get why we'd want this. Can you elaborate? Is this a developer experience issue, to avoid running async tests with `trio`?

Or maybe we could introduce a pytest option (or a mark) to pick the `anyio` backend? I would work as:
- no option, run tests with plain `asyncio` (but through anyio pytest plugin), i.e. uses `AsyncConnection`
- `--anyio-backend=asyncio`, run tests with the `asyncio` backend of `anyio`, i.e. uses `AnyIOConnection` but run the `asyncio` event loop etc.
- `--anyio-backend=trio`, run test with the `trio` backend of `anyio`, i.e. still uses `AnyIOConnection` but with `trio` event loop etc.

Reply:
> The test suite needs anyio's pytest plugin, per first commit (and discussions above) where it replaces pytest-asyncio. But since the plugin comes along with the library, the test suite actually depends on the whole anyio package.

I see, you are right.

> But I'm not sure I get why we'd want this. Can you elaborate?

I just would like to have some test that guarantees that anyio is an optional dependency. I understand that we can't do it anymore, but it would be useful to be able to test without trio.

> Is this a developer experience issue, to avoid running async tests with trio?

Now all the async tests run in two copies, which adds some time to the test run. I think it would be a good idea for the test suite to run with a single backend, yes. I think that would make tests configuration easier, as the backend can be selected globally on tests init.

I think it's a good idea to parametrize the test run from command lines or env vars. I understand we have choices. We can use `--anyio={asyncio|trio|no}` and we also have `--loop={uvloop|default}` (we can change it)... so a whopping 6 choices of async flavous if I'm correct? (as in you can go trio+uvloop too etc?) Maybe we don't need to cover the grid entirely and can trust that certain anyio+uvloop are well tested upstream.

So, I'd make a test run only operate with a given backend selection, and use the CI grid to test things more extensively on push :)

Reply:
Last commit "test: allow to select async backend and implementation" implements this plan.
```
$ pytest --help
[...]
custom options:                                                                                                                         
  --anyio={asyncio,trio}    
                        Use AnyIO implementation of the async API, run tests with specified backend. If unset, use plain asyncio
                        implementation, and run with asyncio backend.
[...]
```